kubeadm: add docs for UpgradeAddonsBeforeControlPlane feature gate (#41606)
kubernetes/website: content/en/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade.md

   [upgrade/kubelet] Now that your control plane is upgraded, please proceed with upgrading your kubelets if you haven't already done so.
   ```

   {{< note >}}

>    Before v1.27 (including v1.27), kubeadm will upgrade the addons (including CoreDNS and kube-proxy) immediately during 

Reviewer: ```suggestion
For versions earlier than v1.28, kubeadm defaulted to a mode that kubeadm will upgrade the addons (including CoreDNS and kube-proxy) immediately during
```

   `kubeadm upgrade apply`, regardless of whether there are other control plane instances that have not been upgraded, which 

>    may cause compatibility problems. Since v1.28, kubeadm will always checks whether all the control plane instances have 

Reviewer: ```suggestion
   may cause compatibility problems. Since v1.28, kubeadm defaults to a mode that checks whether
  all the control plane instances have
```

>    been upgraded before starting to upgrade the addons. You **MUST** perform control plane nodes upgrade sequentially, and 

Author: @neolit123 Because the check is only performed during `kubeadm upgrade apply/node`, when the control plane nodes are upgrading in parallel, there may be a race that all the kubeadm upgrade processes assume that there are unupgraded control plane instances, this will cause the addons to miss performing the upgrade. Complete serial execution is not required, but in order to ensure that the addons can be successfully upgraded during the last cp node upgrade, users should ensure that all the other cp nodes have been upgraded before doing the last cp node upgrade.  In the future, we can separate the addons upgrade into a separate phase and call it after all the cp nodes have been upgraded.

>    been upgraded before starting to upgrade the addons. You **MUST** perform control plane nodes upgrade sequentially, and 

Author: If parallel cp upgrade is necessary, the user needs to make additional checks to see if the addons have been successfully upgraded or always to call `kubeadm upgrade node` on the last cp node again to ensure that the addons can be upgraded successfully. This should be safe because the upgrade operation is idempotent.

>    been upgraded before starting to upgrade the addons. You **MUST** perform control plane nodes upgrade sequentially, and 

Reviewer 2: thanks for the clarification.

   the addons will be upgraded after the last control plane has been upgraded.
   If you want to keep the old upgrade behavior, please enable `UpgradeAddonsBeforeControlPlane` feature gate by 
   `kubeadm upgrade apply --feature-gates=UpgradeAddonsBeforeControlPlane=true`. 

>    But please note that this feature gate has been **deprecated** and will be removed in a future release.

Reviewer: ```suggestion
   The Kubernetes project does not in general recommend enabling this feature gate (you should instead
   change your upgrade process or cluster addons so that you do not need to enable the legacy behavior.
   The `UpgradeAddonsBeforeControlPlane` feature gate will be removed in a future release.
```
